### PR TITLE
Support subtyping in tail calls

### PR DESCRIPTION
**Pull request by tlively**
The tail call spec does not include subtyping because it is based on the
upstream spec, which does not contain subtyping. However, there is no reason
that subtyping shouldn't apply to tail calls like it does for any other call or
return. Update the validator to allow subtyping and avoid a related null pointer
dereference while we're at it.

Do not run the test in with --nominal because it is buggy in that mode.

### CODE REVIEW

File: test/lit/tail-call.wast [hunk 1]
```diff
@@ -0,0 +1,78 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py and should not be edited.
+
+;; Check that tail calls are parsed, validated, and printed correclty
```
Comment (kripken) [inline test/lit/tail-call.wast:3]:
```suggestion
;; Check that tail calls are parsed, validated, and printed correctly
```
(or was it an intentional joke...?)

**Review by kripken — APPROVED** [review]

**Comment by kripken** [comment]
cc @fgmccabe - I'm curious to hear your thoughts and experience on this. My intuition is that it's scary to support subtyping in tail calls, as only when the types are identical then is it really obvious that it's fine to "just" reuse the stack frame. My sense is that this type of thing could be a source of exploits. But I don't have much VM experience here - is my intuition wrong?

File: test/lit/tail-call.wast (hunk 1 above)
Comment (tlively) [reply test/lit/tail-call.wast:3]: Haha no this was an unintentional joke.

**Comment by fgmccabe** [comment]
Thomas is correct. The tail-call is actually return-call which gives an important clue here. When checking types, the return type of the callee should be compared against the return type of the current function. The typing of the arguments to return-call should be the same as for any call.

**Comment by kripken** [comment]
Great, thanks @fgmccabe !

**Pull request merged**